cdap-security/src/main/java/co/cask/cdap/security/impersonation/DefaultImpersonator.java

```diff
@@ -83,13 +83,13 @@ private UserGroupInformation getUGI(NamespacedEntityId entityId,
     }
 
     ImpersonationRequest impersonationRequest = new ImpersonationRequest(entityId, impersonatedOpType);
-    // if the current user is not same as cdap master user then it means we are already impersonating some user
-    // and hence we should not allow another impersonation. See CDAP-8641
-    if (!UserGroupInformation.getCurrentUser().getShortUserName().equals(masterShortUsername)) {
-      LOG.trace("Not impersonating for {} as the call is already impersonated as {}",
-                impersonationRequest, UserGroupInformation.getCurrentUser());
-      return UserGroupInformation.getCurrentUser();
+    UserGroupInformation configuredUGI = ugiProvider.getConfiguredUGI(impersonationRequest).getUGI();
+    // Only cdap master user is allowed to impersonate another user besides itself. See CDAP-8641
+    String currUserShortName = UserGroupInformation.getCurrentUser().getShortUserName();
+    if (currUserShortName.equals(masterShortUsername) || currUserShortName.equals(configuredUGI.getShortUserName())) {
+      return configuredUGI;
     }
-    return ugiProvider.getConfiguredUGI(impersonationRequest).getUGI();
+    throw new UnsupportedOperationException(String.format("User %s is not allowed to impersonate user %s.",
```

Comment:
This shouldn't happen. If it does happen, we should change the code that calls the impersonator, so as to avoid it.

It doesn't seem like an illegal state, but more like the calling code is trying to improperly impersonate. And we don't want to allow that illegal operation.